# Importing Libraries

In [69]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

# Defining names to columns of unlabelled dataset.

In [70]:
names=['id','Clump Thickness','Cell Size','Cell Shape','Marginal Adhesion','Single Epithelial Cell Size','Bare Nuclei','Bland Chromatin','Normal Nucleoli','Mitoses','Class']

In [71]:
df=pd.read_csv("Data.csv",names=names)

In [72]:
df

id  Clump Thickness  Cell Size  Cell Shape  Marginal Adhesion  \
0    1000025                5          1           1                  1   
1    1002945                5          4           4                  5   
2    1015425                3          1           1                  1   
3    1016277                6          8           8                  1   
4    1017023                4          1           1                  3   
..       ...              ...        ...         ...                ...   
694   776715                3          1           1                  1   
695   841769                2          1           1                  1   
696   888820                5         10          10                  3   
697   897471                4          8           6                  4   
698   897471                4          8           8                  5   

     Single Epithelial Cell Size Bare Nuclei  Bland Chromatin  \
0                              2           1                3   
1                              7          10                3   
2                              2           2                3   
3                              3           4                3   
4                              2           1                3   
..                           ...         ...              ...   
694                            3           2                1   
695                            2           1                1   
696                            7           3                8   
697                            3           4               10   
698                            4           5               10   

     Normal Nucleoli  Mitoses  Class  
0                  1        1      2  
1                  2        1      2  
2                  1        1      2  
3                  7        1      2  
4                  1        1      2  
..               ...      ...    ...  
694                1        1      2  
695                1        1      2  
696               10        2      4  
697                6        1      4  
698                4        1      4  

[699 rows x 11 columns]

# Class: (2 for benign, 4 for malignant)

In [73]:
df.describe()

id  Clump Thickness   Cell Size  Cell Shape  \
count  6.990000e+02       699.000000  699.000000  699.000000   
mean   1.071704e+06         4.417740    3.134478    3.207439   
std    6.170957e+05         2.815741    3.051459    2.971913   
min    6.163400e+04         1.000000    1.000000    1.000000   
25%    8.706885e+05         2.000000    1.000000    1.000000   
50%    1.171710e+06         4.000000    1.000000    1.000000   
75%    1.238298e+06         6.000000    5.000000    5.000000   
max    1.345435e+07        10.000000   10.000000   10.000000   

       Marginal Adhesion  Single Epithelial Cell Size  Bland Chromatin  \
count         699.000000                   699.000000       699.000000   
mean            2.806867                     3.216023         3.437768   
std             2.855379                     2.214300         2.438364   
min             1.000000                     1.000000         1.000000   
25%             1.000000                     2.000000         2.000000   
50%             1.000000                     2.000000         3.000000   
75%             4.000000                     4.000000         5.000000   
max            10.000000                    10.000000        10.000000   

       Normal Nucleoli     Mitoses       Class  
count       699.000000  699.000000  699.000000  
mean          2.866953    1.589413    2.689557  
std           3.053634    1.715078    0.951273  
min           1.000000    1.000000    2.000000  
25%           1.000000    1.000000    2.000000  
50%           1.000000    1.000000    2.000000  
75%           4.000000    1.000000    4.000000  
max          10.000000   10.000000    4.000000

In [74]:
df.dtypes

id                              int64
Clump Thickness                 int64
Cell Size                       int64
Cell Shape                      int64
Marginal Adhesion               int64
Single Epithelial Cell Size     int64
Bare Nuclei                    object
Bland Chromatin                 int64
Normal Nucleoli                 int64
Mitoses                         int64
Class                           int64
dtype: object

In [75]:
df.isnull().sum()

id                             0
Clump Thickness                0
Cell Size                      0
Cell Shape                     0
Marginal Adhesion              0
Single Epithelial Cell Size    0
Bare Nuclei                    0
Bland Chromatin                0
Normal Nucleoli                0
Mitoses                        0
Class                          0
dtype: int64

# Hence there is no missing data.

In [76]:
#Now we can drop id column as it is unique for each sample and of no usage in training.

In [77]:
df=df.drop('id',axis='columns')

In [78]:
df

Clump Thickness  Cell Size  Cell Shape  Marginal Adhesion  \
0                  5          1           1                  1   
1                  5          4           4                  5   
2                  3          1           1                  1   
3                  6          8           8                  1   
4                  4          1           1                  3   
..               ...        ...         ...                ...   
694                3          1           1                  1   
695                2          1           1                  1   
696                5         10          10                  3   
697                4          8           6                  4   
698                4          8           8                  5   

     Single Epithelial Cell Size Bare Nuclei  Bland Chromatin  \
0                              2           1                3   
1                              7          10                3   
2                              2           2                3   
3                              3           4                3   
4                              2           1                3   
..                           ...         ...              ...   
694                            3           2                1   
695                            2           1                1   
696                            7           3                8   
697                            3           4               10   
698                            4           5               10   

     Normal Nucleoli  Mitoses  Class  
0                  1        1      2  
1                  2        1      2  
2                  1        1      2  
3                  7        1      2  
4                  1        1      2  
..               ...      ...    ...  
694                1        1      2  
695                1        1      2  
696               10        2      4  
697                6        1      4  
698                4        1      4  

[699 rows x 10 columns]

# Splitting data into X(feature vector) and y(output vector)

In [185]:
X=df.iloc[:,:-1].values

In [186]:
X

array([[5, 1, 1, ..., 3, 1, 1],
       [5, 4, 4, ..., 3, 2, 1],
       [3, 1, 1, ..., 3, 1, 1],
       ...,
       [5, 10, 10, ..., 8, 10, 2],
       [4, 8, 6, ..., 10, 6, 1],
       [4, 8, 8, ..., 10, 4, 1]], dtype=object)

In [187]:
y=df.iloc[:,-1].values

In [188]:
y

array([2, 2, 2, 2, 2, 4, 2, 2, 2, 2, 2, 2, 4, 2, 4, 4, 2, 2, 4, 2, 4, 4,
       2, 4, 2, 4, 2, 2, 2, 2, 2, 2, 4, 2, 2, 2, 4, 2, 4, 4, 2, 4, 4, 4,
       4, 2, 4, 2, 2, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 2, 4, 4, 2, 4,
       2, 4, 4, 2, 2, 4, 2, 4, 4, 2, 2, 2, 2, 2, 2, 2, 2, 2, 4, 4, 4, 4,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 4, 4, 4, 4, 2, 4, 4, 4, 4, 4, 2, 4,
       2, 4, 4, 4, 2, 2, 2, 4, 2, 2, 2, 2, 4, 4, 4, 2, 4, 2, 4, 2, 2, 2,
       4, 2, 2, 2, 2, 2, 2, 2, 2, 2, 4, 2, 2, 2, 4, 2, 2, 4, 2, 4, 4, 2,
       2, 4, 2, 2, 2, 4, 4, 2, 2, 2, 2, 2, 4, 4, 2, 2, 2, 2, 2, 4, 4, 4,
       2, 4, 2, 4, 2, 2, 2, 4, 4, 2, 4, 4, 4, 2, 4, 4, 2, 2, 2, 2, 2, 2,
       2, 2, 4, 4, 2, 2, 2, 4, 4, 2, 2, 2, 4, 4, 2, 4, 4, 4, 2, 2, 4, 2,
       2, 4, 4, 4, 4, 2, 4, 4, 2, 4, 4, 4, 2, 4, 2, 2, 4, 4, 4, 4, 2, 2,
       2, 2, 2, 2, 4, 4, 2, 2, 2, 4, 2, 4, 4, 4, 2, 2, 2, 2, 4, 4, 4, 4,
       4, 2, 4, 4, 4, 2, 4, 2, 4, 4, 2, 2, 2, 2, 2, 4, 2, 2, 4, 4, 4, 4,
       4, 2, 4, 4, 2, 2, 4, 4, 2, 4, 2, 2, 2, 4, 4,

# Splitting data into training and test sets

In [189]:
from sklearn.model_selection import train_test_split

In [190]:
X_train,X_test,y_train,y_test=train_test_split(X,y,test_size=0.25,random_state=0)

In [191]:
print(X_train)

[[8 10 4 ... 8 2 1]
 [3 1 2 ... 2 1 1]
 [8 10 10 ... 4 8 7]
 ...
 [4 1 1 ... 1 1 1]
 [5 1 1 ... 2 1 1]
 [1 1 1 ... 1 1 1]]


In [192]:
print(X_test)

[[4 1 2 ... 1 1 1]
 [4 2 2 ... 2 1 1]
 [6 6 6 ... 7 8 1]
 ...
 [5 3 2 ... 1 1 1]
 [1 1 1 ... 3 1 1]
 [4 1 1 ... 3 2 1]]


In [193]:
print(y_train)

[4 2 4 4 2 4 2 4 2 2 4 4 2 2 2 4 2 4 2 2 2 2 2 2 2 2 2 2 4 2 2 4 2 4 2 4 4
 4 2 4 4 4 2 4 2 2 4 2 2 2 2 2 2 2 2 2 4 2 2 4 2 4 2 4 2 4 2 2 4 2 2 2 2 2
 2 2 2 2 2 4 4 4 2 4 2 4 2 4 2 2 2 2 4 2 4 2 2 2 4 2 4 4 2 2 2 4 4 4 4 2 4
 2 2 2 4 4 4 2 2 4 2 2 2 4 4 2 4 2 4 2 2 2 2 2 4 4 2 2 4 2 4 2 4 4 2 4 2 2
 4 2 2 4 2 2 2 2 2 2 2 4 2 2 4 2 2 2 4 4 4 4 4 2 4 2 2 2 2 2 4 4 2 4 4 4 2
 2 2 4 2 4 4 2 4 2 2 2 2 2 4 2 4 2 2 4 4 2 2 2 4 4 4 2 2 2 2 4 2 4 2 2 2 2
 2 2 2 2 2 4 4 4 2 2 2 4 4 2 2 2 2 2 2 2 4 2 4 2 2 2 4 4 2 4 2 2 2 2 4 2 4
 2 4 2 2 4 2 2 2 2 2 2 2 2 4 4 2 2 2 4 2 4 2 2 2 2 2 4 4 2 2 4 2 2 4 4 2 4
 4 2 2 2 2 2 2 2 2 2 2 2 2 2 2 4 4 2 2 2 2 2 2 4 2 2 2 2 2 2 4 4 2 2 2 2 2
 2 4 4 2 2 4 2 2 2 2 4 2 2 4 4 2 2 4 2 2 2 4 2 2 4 2 4 2 2 4 4 4 2 2 2 4 2
 2 2 2 4 2 2 2 2 4 2 4 2 4 4 2 2 2 4 4 4 2 2 2 2 4 2 2 4 2 4 2 4 2 2 4 2 2
 2 2 2 2 4 2 2 2 4 4 4 4 4 2 2 2 2 4 4 4 4 2 2 4 2 2 2 2 2 4 2 2 4 2 4 2 2
 2 2 2 2 4 4 2 2 4 2 2 2 2 2 2 4 4 4 4 2 2 2 2 2 2 4 2 2 2 2 4 2 2 2 4 4 2
 4 4 4 2 4 2 2 4 2 2 2 2 

In [194]:
print(y_test)

[2 2 2 2 4 2 4 2 4 4 2 2 4 4 4 2 2 4 4 2 4 4 2 2 2 4 2 2 4 2 2 2 2 2 2 2 2
 4 2 2 2 2 2 2 4 4 2 4 2 4 4 2 2 4 2 2 2 2 2 2 4 2 2 4 4 4 4 2 2 4 2 2 4 4
 2 2 2 2 4 2 2 2 4 2 2 2 4 2 4 4 2 4 2 4 2 2 2 4 2 4 4 2 2 2 4 2 2 2 2 2 4
 4 4 2 2 2 2 2 4 4 4 4 2 4 2 2 4 4 4 4 4 2 2 4 4 2 2 4 2 2 4 4 2 2 2 2 2 4
 2 2 2 4 4 2 2 2 2 2 2 2 4 2 4 2 2 2 2 2 2 2 4 2 2 2 2]


# Feature Scaling

In [195]:
df=df.replace('?',-99999)

In [196]:
df.head()

Clump Thickness Cell Size Cell Shape Marginal Adhesion  \
0               5         1          1                 1   
1               5         4          4                 5   
2               3         1          1                 1   
3               6         8          8                 1   
4               4         1          1                 3   

  Single Epithelial Cell Size Bare Nuclei Bland Chromatin Normal Nucleoli  \
0                           2           1               3               1   
1                           7          10               3               2   
2                           2           2               3               1   
3                           3           4               3               7   
4                           2           1               3               1   

  Mitoses  Class  
0       1      2  
1       1      2  
2       1      2  
3       1      2  
4       1      2

In [197]:
df=df.replace('?',-9999)

In [198]:
from sklearn.preprocessing import StandardScaler
sc = StandardScaler()

In [199]:
X_train.reshape(-1,1)

array([[8],
       [10],
       [4],
       ...,
       [1],
       [1],
       [1]], dtype=object)

In [200]:
df

Clump Thickness Cell Size Cell Shape Marginal Adhesion  \
0                 5         1          1                 1   
1                 5         4          4                 5   
2                 3         1          1                 1   
3                 6         8          8                 1   
4                 4         1          1                 3   
..              ...       ...        ...               ...   
694               3         1          1                 1   
695               2         1          1                 1   
696               5        10         10                 3   
697               4         8          6                 4   
698               4         8          8                 5   

    Single Epithelial Cell Size Bare Nuclei Bland Chromatin Normal Nucleoli  \
0                             2           1               3               1   
1                             7          10               3               2   
2                             2           2               3               1   
3                             3           4               3               7   
4                             2           1               3               1   
..                          ...         ...             ...             ...   
694                           3           2               1               1   
695                           2           1               1               1   
696                           7           3               8              10   
697                           3           4              10               6   
698                           4           5              10               4   

    Mitoses  Class  
0         1      2  
1         1      2  
2         1      2  
3         1      2  
4         1      2  
..      ...    ...  
694       1      2  
695       1      2  
696       2      4  
697       1      4  
698       1      4  

[699 rows x 10 columns]

In [201]:
from sklearn.preprocessing import LabelEncoder
le = LabelEncoder()
y = le.fit_transform(y)
print(y)

[0 0 0 0 0 1 0 0 0 0 0 0 1 0 1 1 0 0 1 0 1 1 0 1 0 1 0 0 0 0 0 0 1 0 0 0 1
 0 1 1 0 1 1 1 1 0 1 0 0 1 1 1 1 1 1 1 1 1 1 1 1 0 1 1 0 1 0 1 1 0 0 1 0 1
 1 0 0 0 0 0 0 0 0 0 1 1 1 1 0 0 0 0 0 0 0 0 0 0 1 1 1 1 0 1 1 1 1 1 0 1 0
 1 1 1 0 0 0 1 0 0 0 0 1 1 1 0 1 0 1 0 0 0 1 0 0 0 0 0 0 0 0 0 1 0 0 0 1 0
 0 1 0 1 1 0 0 1 0 0 0 1 1 0 0 0 0 0 1 1 0 0 0 0 0 1 1 1 0 1 0 1 0 0 0 1 1
 0 1 1 1 0 1 1 0 0 0 0 0 0 0 0 1 1 0 0 0 1 1 0 0 0 1 1 0 1 1 1 0 0 1 0 0 1
 1 1 1 0 1 1 0 1 1 1 0 1 0 0 1 1 1 1 0 0 0 0 0 0 1 1 0 0 0 1 0 1 1 1 0 0 0
 0 1 1 1 1 1 0 1 1 1 0 1 0 1 1 0 0 0 0 0 1 0 0 1 1 1 1 1 0 1 1 0 0 1 1 0 1
 0 0 0 1 1 0 1 0 1 1 0 0 1 0 0 0 1 0 0 0 1 1 0 0 1 0 0 1 0 0 1 0 1 1 1 0 0
 1 1 0 1 0 0 1 1 0 0 0 1 0 0 0 1 1 0 0 0 1 0 0 1 1 1 1 1 1 0 0 0 0 1 1 0 0
 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 1 0 0 0 0 1 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0
 0 0 0 0 0 1 0 1 0 1 0 0 0 0 1 0 0 0 1 0 1 0 0 0 0 0 0 0 1 1 0 0 0 1 0 0 0
 0 0 0 0 0 1 0 0 0 1 0 1 1 1 0 0 0 0 0 0 0 1 1 1 0 0 0 0 0 0 0 0 0 0 0 1 0
 0 1 1 0 0 0 1 1 1 0 1 0 

In [202]:
from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import OneHotEncoder
ct = ColumnTransformer(transformers=[('encoder', OneHotEncoder(), [0])], remainder='passthrough')
X = np.array(ct.fit_transform(X))
print(X)

[[0.0 0.0 0.0 ... 3 1 1]
 [0.0 0.0 0.0 ... 3 2 1]
 [0.0 0.0 1.0 ... 3 1 1]
 ...
 [0.0 0.0 0.0 ... 8 10 2]
 [0.0 0.0 0.0 ... 10 6 1]
 [0.0 0.0 0.0 ... 10 4 1]]


In [203]:
df['Clump Thickness'] = df['Clump Thickness'].astype('category')
df['Cell Size'] = df['Cell Size'].astype('category')
df['Cell Shape'] = df['Cell Shape'].astype('category')
df['Marginal Adhesion']=df['Marginal Adhesion'].astype('category')
df['Single Epithelial Cell Size'] = df['Single Epithelial Cell Size'].astype('category')
df['Bare Nuclei'] = df['Bare Nuclei'].astype('category')
df['Bland Chromatin'] = df['Bland Chromatin'].astype('category')
df['Normal Nucleoli'] = df['Normal Nucleoli'].astype('category')
df['Mitoses'] = df['Mitoses'].astype('category')

In [204]:
df.dtypes

Clump Thickness                category
Cell Size                      category
Cell Shape                     category
Marginal Adhesion              category
Single Epithelial Cell Size    category
Bare Nuclei                    category
Bland Chromatin                category
Normal Nucleoli                category
Mitoses                        category
Class                             int64
dtype: object

In [205]:
X_train[:, 1:] = sc.fit_transform(X_train[:, 1:])
X_test[:, 1:] = sc.transform(X_test[:, 1:])

In [206]:
print(X_train)

[[8 2.212792165478631 0.2679068663921393 ... 1.8758447129125353
  -0.2955217297362437 -0.33390732320588634]
 [3 -0.6859655712983757 -0.3958434787117898 ... -0.5845874011454584
  -0.6154667429217636 -0.33390732320588634]
 [8 2.212792165478631 2.2591579017039267 ... 0.23555663687387282
  1.6241483493768762 3.2861086221710334]
 ...
 [4 -0.6859655712983757 -0.7277186512637543 ... -0.994659420155124
  -0.6154667429217636 -0.33390732320588634]
 [5 -0.6859655712983757 -0.7277186512637543 ... -0.5845874011454584
  -0.6154667429217636 -0.33390732320588634]
 [1 -0.6859655712983757 -0.7277186512637543 ... -0.994659420155124
  -0.6154667429217636 -0.33390732320588634]]


In [207]:
print(X_test)

[[4 -0.6859655712983757 -0.3958434787117898 ... -0.994659420155124
  -0.6154667429217636 -0.33390732320588634]
 [4 -0.36388137832315265 -0.3958434787117898 ... -0.5845874011454584
  -0.6154667429217636 -0.33390732320588634]
 [6 0.9244553935777391 0.9316572114960684 ... 1.4657726939028697
  1.6241483493768762 -0.33390732320588634]
 ...
 [5 -0.04179718534792973 -0.3958434787117898 ... -0.994659420155124
  -0.6154667429217636 -0.33390732320588634]
 [1 -0.6859655712983757 -0.7277186512637543 ... -0.1745153821357928
  -0.6154667429217636 -0.33390732320588634]
 [4 -0.6859655712983757 -0.7277186512637543 ... -0.1745153821357928
  -0.2955217297362437 -0.33390732320588634]]


# Now applying KNN to the endoded data.

In [209]:
from sklearn.neighbors import KNeighborsClassifier
classifier = KNeighborsClassifier(n_neighbors = 5, metric = 'minkowski', p = 2)
classifier.fit(X_train, y_train)

KNeighborsClassifier()

# Predicting a result from dataframe data as input

In [236]:
df

Clump Thickness Cell Size Cell Shape Marginal Adhesion  \
0                 5         1          1                 1   
1                 5         4          4                 5   
2                 3         1          1                 1   
3                 6         8          8                 1   
4                 4         1          1                 3   
..              ...       ...        ...               ...   
694               3         1          1                 1   
695               2         1          1                 1   
696               5        10         10                 3   
697               4         8          6                 4   
698               4         8          8                 5   

    Single Epithelial Cell Size Bare Nuclei Bland Chromatin Normal Nucleoli  \
0                             2           1               3               1   
1                             7          10               3               2   
2                             2           2               3               1   
3                             3           4               3               7   
4                             2           1               3               1   
..                          ...         ...             ...             ...   
694                           3           2               1               1   
695                           2           1               1               1   
696                           7           3               8              10   
697                           3           4              10               6   
698                           4           5              10               4   

    Mitoses  Class  
0         1      2  
1         1      2  
2         1      2  
3         1      2  
4         1      2  
..      ...    ...  
694       1      2  
695       1      2  
696       2      4  
697       1      4  
698       1      4  

[699 rows x 10 columns]

In [239]:
X_train[1,:]

array([3, -0.6859655712983757, -0.3958434787117898, -0.6330752467929711,
       -0.5581603638337332, 0.15934179796471362, -0.5845874011454584,
       -0.6154667429217636, -0.33390732320588634], dtype=object)

In [241]:
y_pred = classifier.predict(X_test)
print(np.concatenate((y_pred.reshape(len(y_pred),1), y_test.reshape(len(y_test),1)),1))

[[2 2]
 [2 2]
 [2 2]
 [2 2]
 [4 4]
 [2 2]
 [4 4]
 [2 2]
 [4 4]
 [4 4]
 [2 2]
 [2 2]
 [4 4]
 [4 4]
 [4 4]
 [2 2]
 [2 2]
 [4 4]
 [4 4]
 [2 2]
 [4 4]
 [4 4]
 [2 2]
 [2 2]
 [2 2]
 [4 4]
 [2 2]
 [2 2]
 [4 4]
 [4 2]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [4 4]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [4 4]
 [4 4]
 [2 2]
 [4 4]
 [2 2]
 [4 4]
 [4 4]
 [2 2]
 [2 2]
 [2 4]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [4 4]
 [2 2]
 [2 2]
 [4 4]
 [4 4]
 [4 4]
 [4 4]
 [2 2]
 [2 2]
 [4 4]
 [2 2]
 [2 2]
 [4 4]
 [4 4]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [4 4]
 [2 2]
 [2 2]
 [2 2]
 [4 4]
 [2 2]
 [2 2]
 [2 2]
 [4 4]
 [2 2]
 [4 4]
 [4 4]
 [2 2]
 [2 4]
 [4 2]
 [4 4]
 [2 2]
 [2 2]
 [2 2]
 [4 4]
 [2 2]
 [4 4]
 [4 4]
 [2 2]
 [2 2]
 [2 2]
 [4 4]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [4 4]
 [4 4]
 [4 4]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [4 4]
 [2 4]
 [4 4]
 [4 4]
 [2 2]
 [4 4]
 [2 2]
 [2 2]
 [4 4]
 [4 4]
 [4 4]
 [4 4]
 [4 4]
 [2 2]
 [2 2]
 [4 4]
 [4 4]
 [2 2]
 [2 2]
 [4 4]
 [2 2]
 [2 2]
 [4 4]
 [4 4]
 [2 2]

# Checking for the accuracy for KNN model.

In [243]:
from sklearn.metrics import confusion_matrix, accuracy_score
cm = confusion_matrix(y_test, y_pred)

In [244]:
print(cm)

[[109   3]
 [  5  58]]


In [245]:
print(accuracy_score(y_test,y_pred))

0.9542857142857143


# Thus by using KNN we get an accuracy score of 95.43%.

# Now using SVM 

In [248]:
from sklearn.svm import SVC

In [249]:
classifier=SVC(kernel='linear',random_state=0)


In [250]:
classifier.fit(X_train,y_train)

SVC(kernel='linear', random_state=0)

In [251]:
y_pred=classifier.predict(X_test)
print(np.concatenate((y_pred.reshape(len(y_pred),1), y_test.reshape(len(y_test),1)),1))

[[2 2]
 [2 2]
 [4 2]
 [2 2]
 [4 4]
 [2 2]
 [4 4]
 [2 2]
 [4 4]
 [4 4]
 [2 2]
 [2 2]
 [4 4]
 [4 4]
 [4 4]
 [2 2]
 [2 2]
 [4 4]
 [4 4]
 [2 2]
 [4 4]
 [4 4]
 [2 2]
 [2 2]
 [2 2]
 [4 4]
 [2 2]
 [2 2]
 [4 4]
 [4 2]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [4 4]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [4 4]
 [4 4]
 [2 2]
 [4 4]
 [2 2]
 [4 4]
 [4 4]
 [2 2]
 [2 2]
 [4 4]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [4 4]
 [2 2]
 [2 2]
 [4 4]
 [4 4]
 [4 4]
 [4 4]
 [2 2]
 [2 2]
 [4 4]
 [2 2]
 [2 2]
 [4 4]
 [4 4]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [4 4]
 [2 2]
 [2 2]
 [2 2]
 [4 4]
 [2 2]
 [2 2]
 [2 2]
 [4 4]
 [2 2]
 [4 4]
 [4 4]
 [2 2]
 [2 4]
 [4 2]
 [4 4]
 [2 2]
 [2 2]
 [2 2]
 [4 4]
 [2 2]
 [4 4]
 [4 4]
 [2 2]
 [2 2]
 [2 2]
 [4 4]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [4 4]
 [4 4]
 [4 4]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [4 4]
 [4 4]
 [4 4]
 [4 4]
 [2 2]
 [4 4]
 [2 2]
 [2 2]
 [4 4]
 [4 4]
 [4 4]
 [4 4]
 [4 4]
 [2 2]
 [2 2]
 [4 4]
 [4 4]
 [2 2]
 [2 2]
 [4 4]
 [2 2]
 [2 2]
 [4 4]
 [4 4]
 [2 2]

In [252]:
from sklearn.metrics import confusion_matrix, accuracy_score
cm = confusion_matrix(y_test, y_pred)
print(cm)
accuracy_score(y_test, y_pred)

[[108   4]
 [  3  60]]


0.96

# Using SVM the accuracy is improved a bit-96%.

# Now we will try using radial basis function kernel SVM and check for any improvement in model with same dataset and same features.

In [254]:
from sklearn.svm import SVC
classifier_kernel=SVC(kernel='rbf',random_state=0)
classifier_kernel.fit(X_train,y_train)

SVC(random_state=0)

In [255]:
y_pred=classifier_kernel.predict(X_test)
print(np.concatenate((y_pred.reshape(len(y_pred),1), y_test.reshape(len(y_test),1)),1))

[[2 2]
 [2 2]
 [4 2]
 [2 2]
 [4 4]
 [2 2]
 [4 4]
 [2 2]
 [4 4]
 [4 4]
 [2 2]
 [2 2]
 [4 4]
 [4 4]
 [4 4]
 [2 2]
 [2 2]
 [4 4]
 [4 4]
 [2 2]
 [4 4]
 [4 4]
 [2 2]
 [2 2]
 [2 2]
 [4 4]
 [2 2]
 [2 2]
 [4 4]
 [4 2]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [4 4]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [4 4]
 [4 4]
 [2 2]
 [4 4]
 [2 2]
 [4 4]
 [4 4]
 [2 2]
 [2 2]
 [2 4]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [4 4]
 [2 2]
 [2 2]
 [4 4]
 [4 4]
 [4 4]
 [4 4]
 [2 2]
 [2 2]
 [4 4]
 [2 2]
 [2 2]
 [4 4]
 [4 4]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [4 4]
 [2 2]
 [2 2]
 [2 2]
 [4 4]
 [2 2]
 [2 2]
 [2 2]
 [4 4]
 [2 2]
 [4 4]
 [4 4]
 [2 2]
 [2 4]
 [4 2]
 [4 4]
 [2 2]
 [2 2]
 [2 2]
 [4 4]
 [2 2]
 [4 4]
 [4 4]
 [2 2]
 [2 2]
 [2 2]
 [4 4]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [4 4]
 [4 4]
 [4 4]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [4 4]
 [2 4]
 [4 4]
 [4 4]
 [2 2]
 [4 4]
 [2 2]
 [2 2]
 [4 4]
 [4 4]
 [4 4]
 [4 4]
 [4 4]
 [2 2]
 [2 2]
 [4 4]
 [4 4]
 [2 2]
 [2 2]
 [4 4]
 [2 2]
 [2 2]
 [4 4]
 [4 4]
 [2 2]

In [256]:
from sklearn.metrics import confusion_matrix, accuracy_score
cm = confusion_matrix(y_test, y_pred)
print(cm)
accuracy_score(y_test, y_pred)

[[108   4]
 [  5  58]]


0.9485714285714286

# The efficiency is not improved.So using linear kernel gives better results than using rbf kernel.

# Naive Bayes.

In [259]:
from sklearn.naive_bayes import GaussianNB
classifier = GaussianNB()
classifier.fit(X_train, y_train)

GaussianNB()

# Using Guassian Kernel

In [260]:
y_pred = classifier.predict(X_test)
print(np.concatenate((y_pred.reshape(len(y_pred),1), y_test.reshape(len(y_test),1)),1))

[[2 2]
 [2 2]
 [4 2]
 [2 2]
 [4 4]
 [2 2]
 [4 4]
 [2 2]
 [4 4]
 [4 4]
 [2 2]
 [2 2]
 [4 4]
 [4 4]
 [4 4]
 [2 2]
 [2 2]
 [4 4]
 [4 4]
 [2 2]
 [4 4]
 [4 4]
 [2 2]
 [2 2]
 [2 2]
 [4 4]
 [2 2]
 [2 2]
 [4 4]
 [4 2]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [4 2]
 [2 2]
 [4 4]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [4 4]
 [4 4]
 [2 2]
 [4 4]
 [2 2]
 [4 4]
 [4 4]
 [2 2]
 [2 2]
 [4 4]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [4 2]
 [2 2]
 [4 4]
 [2 2]
 [2 2]
 [4 4]
 [4 4]
 [4 4]
 [4 4]
 [2 2]
 [2 2]
 [4 4]
 [2 2]
 [2 2]
 [4 4]
 [4 4]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [4 4]
 [2 2]
 [2 2]
 [2 2]
 [4 4]
 [2 2]
 [2 2]
 [2 2]
 [4 4]
 [2 2]
 [4 4]
 [4 4]
 [2 2]
 [2 4]
 [4 2]
 [4 4]
 [2 2]
 [2 2]
 [2 2]
 [4 4]
 [2 2]
 [4 4]
 [4 4]
 [2 2]
 [2 2]
 [2 2]
 [4 4]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [4 4]
 [4 4]
 [4 4]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [4 4]
 [4 4]
 [4 4]
 [4 4]
 [2 2]
 [4 4]
 [2 2]
 [2 2]
 [4 4]
 [4 4]
 [4 4]
 [4 4]
 [4 4]
 [2 2]
 [2 2]
 [4 4]
 [4 4]
 [2 2]
 [2 2]
 [4 4]
 [2 2]
 [2 2]
 [4 4]
 [4 4]
 [2 2]

In [261]:
from sklearn.metrics import confusion_matrix, accuracy_score
cm = confusion_matrix(y_test, y_pred)
print(cm)
accuracy_score(y_test, y_pred)


[[106   6]
 [  2  61]]


0.9542857142857143

# Efficiency-95.43% 

# Using Decision tree classification

In [263]:
from sklearn.tree import DecisionTreeClassifier
classifier = DecisionTreeClassifier(criterion = 'entropy', random_state = 0)
classifier.fit(X_train, y_train)

DecisionTreeClassifier(criterion='entropy', random_state=0)

In [264]:
y_pred = classifier.predict(X_test)
print(np.concatenate((y_pred.reshape(len(y_pred),1), y_test.reshape(len(y_test),1)),1))

[[2 2]
 [2 2]
 [4 2]
 [2 2]
 [4 4]
 [2 2]
 [4 4]
 [2 2]
 [4 4]
 [2 4]
 [2 2]
 [2 2]
 [4 4]
 [2 4]
 [4 4]
 [2 2]
 [2 2]
 [4 4]
 [4 4]
 [2 2]
 [4 4]
 [4 4]
 [2 2]
 [2 2]
 [2 2]
 [4 4]
 [2 2]
 [2 2]
 [4 4]
 [4 2]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [4 4]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [4 4]
 [4 4]
 [2 2]
 [4 4]
 [2 2]
 [4 4]
 [4 4]
 [2 2]
 [2 2]
 [4 4]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [4 4]
 [2 2]
 [2 2]
 [4 4]
 [4 4]
 [4 4]
 [4 4]
 [2 2]
 [2 2]
 [2 4]
 [2 2]
 [2 2]
 [4 4]
 [4 4]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [4 4]
 [2 2]
 [2 2]
 [2 2]
 [4 4]
 [2 2]
 [2 2]
 [2 2]
 [4 4]
 [2 2]
 [4 4]
 [4 4]
 [2 2]
 [4 4]
 [4 2]
 [4 4]
 [2 2]
 [2 2]
 [2 2]
 [4 4]
 [2 2]
 [4 4]
 [4 4]
 [2 2]
 [2 2]
 [2 2]
 [4 4]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [4 4]
 [4 4]
 [4 4]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [4 4]
 [4 4]
 [4 4]
 [4 4]
 [2 2]
 [4 4]
 [2 2]
 [2 2]
 [4 4]
 [2 4]
 [4 4]
 [4 4]
 [4 4]
 [2 2]
 [2 2]
 [4 4]
 [4 4]
 [2 2]
 [2 2]
 [4 4]
 [2 2]
 [2 2]
 [4 4]
 [4 4]
 [2 2]

In [266]:
from sklearn.metrics import confusion_matrix, accuracy_score
cm = confusion_matrix(y_test, y_pred)
print(cm)
accuracy_score(y_test, y_pred)

[[107   5]
 [  6  57]]


0.9371428571428572

# Using Random Forest Classification

In [267]:
from sklearn.ensemble import RandomForestClassifier
classifier = RandomForestClassifier(n_estimators = 10, criterion = 'entropy', random_state = 0)
classifier.fit(X_train, y_train)

RandomForestClassifier(criterion='entropy', n_estimators=10, random_state=0)

In [268]:
y_pred = classifier.predict(X_test)
print(np.concatenate((y_pred.reshape(len(y_pred),1), y_test.reshape(len(y_test),1)),1))

[[2 2]
 [2 2]
 [4 2]
 [2 2]
 [4 4]
 [2 2]
 [4 4]
 [2 2]
 [4 4]
 [2 4]
 [2 2]
 [2 2]
 [4 4]
 [4 4]
 [4 4]
 [2 2]
 [2 2]
 [4 4]
 [4 4]
 [2 2]
 [4 4]
 [4 4]
 [2 2]
 [2 2]
 [2 2]
 [4 4]
 [2 2]
 [2 2]
 [4 4]
 [4 2]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [4 4]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [4 4]
 [4 4]
 [2 2]
 [4 4]
 [2 2]
 [4 4]
 [4 4]
 [2 2]
 [2 2]
 [4 4]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [4 4]
 [2 2]
 [2 2]
 [4 4]
 [4 4]
 [4 4]
 [4 4]
 [2 2]
 [2 2]
 [4 4]
 [2 2]
 [2 2]
 [4 4]
 [4 4]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [4 4]
 [2 2]
 [2 2]
 [2 2]
 [4 4]
 [2 2]
 [2 2]
 [2 2]
 [4 4]
 [2 2]
 [4 4]
 [4 4]
 [2 2]
 [2 4]
 [2 2]
 [4 4]
 [2 2]
 [2 2]
 [2 2]
 [4 4]
 [2 2]
 [4 4]
 [4 4]
 [2 2]
 [2 2]
 [2 2]
 [4 4]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [4 4]
 [4 4]
 [4 4]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [4 4]
 [4 4]
 [4 4]
 [4 4]
 [2 2]
 [4 4]
 [2 2]
 [2 2]
 [4 4]
 [4 4]
 [4 4]
 [4 4]
 [4 4]
 [2 2]
 [2 2]
 [4 4]
 [4 4]
 [2 2]
 [2 2]
 [4 4]
 [2 2]
 [2 2]
 [4 4]
 [4 4]
 [2 2]

In [269]:
from sklearn.metrics import confusion_matrix, accuracy_score
cm = confusion_matrix(y_test, y_pred)
print(cm)
accuracy_score(y_test, y_pred)

[[109   3]
 [  3  60]]


0.9657142857142857

# Efficiency is improved with the same parameters(96.57%)

Thus we can conclude:
Using different classifiers-KNN,SVM(with linear,rbf kernels),Naive Bayes(Guassian kernel),Decision tree and random forest classification we have got the best accuracy with random forest classifier model-96.57%.So using this classifier can be a positive insight in medical science for prediction of breast cancer.